src/webob/acceptparse.py

```diff
@@ -268,6 +678,1197 @@ def _match(self, mask, item):
+
+        if isinstance(
+            other, (AcceptLanguageNoHeader, AcceptLanguageInvalidHeader)
+        ):
+            return AcceptLanguageNoHeader()
+
+        return self._add_instance_and_non_accept_language_type(
+            instance=self, other=other,
+        )
+
+    def __radd__(self, other):
+        """
+        Add to header, creating a new header object.
+
+        See the docstring for :meth:`AcceptLanguageValidHeader.__add__`.
+        """
+        return self._add_instance_and_non_accept_language_type(
+            instance=self, other=other, instance_on_the_right=True,
+        )
+
+    def __repr__(self):
+        return '<{}>'.format(self.__class__.__name__)
+        # We do not display the header_value, as it is untrusted input. The
+        # header_value could always be easily obtained from the .header_value
+        # property.
+
+    def __str__(self):
+        """Return the ``str`` ``'<invalid header value>'``."""
+        return '<invalid header value>'
+
+    def _add_instance_and_non_accept_language_type(
+        self, instance, other, instance_on_the_right=False,
+    ):
+        if not other:
+            return AcceptLanguageNoHeader()
+
+        other_header_value = self.python_value_to_header_str(value=other)
+
+        try:
+            return AcceptLanguageValidHeader(header_value=other_header_value)
+        except ValueError:  # invalid header value
+            return AcceptLanguageNoHeader()
+
+
+def create_accept_language_header(header_value):
+    """
+    Create an object representing the ``Accept-Language`` header in a request.
+
+    :param header_value: (``str``) header value
+    :return: If `header_value` is ``None``, an :class:`AcceptLanguageNoHeader`
+             instance.
+
+             | If `header_value` is a valid ``Accept-Language`` header, an
+               :class:`AcceptLanguageValidHeader` instance.
+
+             | If `header_value` is an invalid ``Accept-Language`` header, an
+               :class:`AcceptLanguageInvalidHeader` instance.
+    """
+    if header_value is None:
+        return AcceptLanguageNoHeader()
+    try:
+        return AcceptLanguageValidHeader(header_value=header_value)
+    except ValueError:
+        return AcceptLanguageInvalidHeader(header_value=header_value)
+
+
+def accept_language_property():
+    doc = """
+        Property representing the ``Accept-Language`` header.
+
+        (:rfc:`RFC 7231, section 5.3.5 <7231#section-5.3.5>`)
+
+        The header value in the request environ is parsed and a new object
+        representing the header is created every time we *get* the value of the
+        property. (*set* and *del* change the header value in the request
+        environ, and do not involve parsing.)'
+    """
+
+    ENVIRON_KEY = 'HTTP_ACCEPT_LANGUAGE'
+
+    def fget(request):
+        """Get an object representing the header in the request."""
+        return create_accept_language_header(
```

Comment:
It doesn't assign a new `AcceptLanguage` when we fset though: fset only changes the header value in the environ, so that the next time we fget, a brand new `AcceptLanguage*` object is created using that new header value in the environ. This is the same as how it current works in `accept_property`.

I can't quite understand the second paragraph? In terms of overhead, we talked a while back about whether we should cache the object, as a new object was being created on every attribute read, and you said that it wasn't a concern at that point &mdash; is that related to what you're asking here?

With the `__add__` functions,

    request.accept_language += 'nl-NL;q=0.7'

is the same as

    request.accept_language = request.accept_language + 'nl-NL;q=0.7'

so it works :)